### PR TITLE
Expose Z3_solver_from_string and Z3_optimize_from_string

### PR DESCRIPTION
**Pull request by verpeteren**

### CODE REVIEW

File: z3/src/solver.rs [hunk 1]
```diff
@@ -62,6 +62,22 @@ impl<'ctx> Solver<'ctx> {
         }
     }
 
+    /// Parse an SMT-LIB2 string with assertions, soft constraints and optimization objectives.
+    /// Add the parsed constraints and objectives to a new created solver context.
+    pub fn new_from_smtlib2(ctx: &'ctx Context, source_string: String) -> Solver<'ctx> {
```
Comment (waywardmonkeys) [inline z3/src/solver.rs:67]: Can this be `&str` instead of `String`?

File: z3/src/optimize.rs [hunk 2]
```diff
@@ -30,6 +30,22 @@ impl<'ctx> Optimize<'ctx> {
         }
     }
 
+    /// Parse an SMT-LIB2 string with assertions, soft constraints and optimization objectives.
+    /// Add the parsed constraints and objectives to a new created optimization context.
+    pub fn new_from_smtlib2(ctx: &'ctx Context, source_string: String) -> Optimize<'ctx> {
```
Comment (waywardmonkeys) [inline z3/src/optimize.rs:35]: As below, could this be `&str` rather than `String`?

File: z3/src/solver.rs [hunk 3]
```diff
@@ -62,6 +62,22 @@ impl<'ctx> Solver<'ctx> {
...
+        let source_cstring = CString::new(source_string).unwrap();
+        Solver {
+            ctx,
+            z3_slv: unsafe {
+                let _guard = Z3_MUTEX.lock().unwrap();
+                let opt = Z3_mk_solver(ctx.z3_ctx);
+                Z3_solver_from_string(ctx.z3_ctx, opt, source_cstring.as_ptr());
```
Comment (waywardmonkeys) [inline z3/src/solver.rs:74]: This can now use `Solver::wrap` for parts of this logic.

File: z3/src/optimize.rs [hunk 4]
```diff
@@ -30,6 +30,22 @@ impl<'ctx> Optimize<'ctx> {
...
+        let source_cstring = CString::new(source_string).unwrap();
+        Optimize {
+            ctx,
+            z3_opt: unsafe {
+                let _guard = Z3_MUTEX.lock().unwrap();
+                let opt = Z3_mk_optimize(ctx.z3_ctx);
+                Z3_optimize_inc_ref(ctx.z3_ctx, opt);
+                Z3_optimize_from_string(ctx.z3_ctx, opt, source_cstring.as_ptr());
```
Comment (waywardmonkeys) [inline z3/src/optimize.rs:43]: This can now use `Optimize::wrap` for parts of this logic.

**Comment by Pat-Lafon** [comment]
@waywardmonkeys fyi this pr can be closed as completed in #226

**Comment by waywardmonkeys** [comment]
Okay! Thanks for the work!

**Pull request closed without merging**